arrows/core/generate_texture.h

```diff
@@ -0,0 +1,441 @@
+    }
+  };
+  for (int n = 0; n < nb_iter; ++n)
+  {
+    // horizontal dilate
+    for (unsigned int y = 0; y < height; ++y)
+    {
+      for (unsigned int x = 0; x < width - 1; ++x)
+      {
+        char d = mask(x + 1, y) - mask(x, y);
+        if (d == 1)
+        {
+          mask(x, y) = 1;
+          copy_pixel(x, y, x + 1, y);
+        }
+        else if (d == -1)
+        {
+          mask(x + 1, y) = 1;
+          copy_pixel(x + 1, y, x, y);
+          ++x;
+        }
+      }
+    }
+    // vertical dilate
+    for (unsigned int x = 0; x < width; ++x)
+    {
+      for (unsigned int y = 0; y < height - 1; ++y)
+      {
+        char d = mask(x, y + 1) - mask(x, y);
+        if (d == 1)
+        {
+          mask(x, y) = 1;
+          copy_pixel(x, y, x, y + 1);
+        }
+        else if (d == -1)
+        {
+          mask(x, y + 1) = 1;
+          copy_pixel(x, y + 1, x, y);
+          ++y;
+        }
+      }
+    }
+  }
+}
+
+
+/// This function generates a texture from a set of images and maps it on the mesh
+/**
+ * \param mesh [in/out] the mesh to texture.
+ * \param cameras [in] a list of cameras perspective
+ * \param images [in] a list of images
+ * \param resolution [in] resolution of the texture (mesh unit/pixel)
+ */
+template <class T>
+vital::image_container_sptr
+generate_texture(vital::mesh_sptr mesh, std::vector<vital::camera_perspective_sptr> const& cameras,
+                 std::vector<vital::image> const& images, double resolution)
+{
+  if (mesh->faces().regularity() != 3)
+  {
+    LOG_ERROR(vital::get_logger("arrows.core.generate_texture" ), "The mesh has to be triangular.");
+    return nullptr;
+  }
+
+  kwiver::vital::mesh_vertex_array<3>& vertices = dynamic_cast< kwiver::vital::mesh_vertex_array<3>& >(mesh->vertices());
+  auto const& triangles = static_cast< kwiver::vital::mesh_regular_face_array<3> const& >(mesh->faces());
+
+  // Unwrap the mesh
+  if (mesh->has_tex_coords() == 0)
+  {
+    uv_unwrap_mesh unwrap;
+    vital::config_block_sptr config = unwrap.get_configuration();
+    unwrap.set_configuration(config);
+    unwrap.unwrap(mesh);
+  }
+
+  auto tcoords = mesh->tex_coords();
+  // Rescale tcoords to real pixel values
+  size_t scale = 1;
+  for (unsigned int f = 0; f < mesh->num_faces(); ++f)
+  {
+    vital::matrix_3x3d points_2d_h;
+    points_2d_h << tcoords[f * 3 + 0], tcoords[f * 3 + 1], tcoords[f * 3 + 2], 1, 1, 1;
+    double area_2d = points_2d_h.determinant();
+    auto const& v1 = vertices[triangles(f, 0)];
+    auto const& v2 = vertices[triangles(f, 1)];
+    auto const& v3 = vertices[triangles(f, 2)];
+    vital::vector_3d a3 = v2 - v1;
+    vital::vector_3d b3 = v3 - v1;
+    double area_3d = a3.cross(b3).norm();
+
+    if (area_2d > 0 && area_3d > 0 && !std::isinf(area_2d) && !std::isinf(area_3d))
+    {
+      scale = static_cast<size_t>(std::ceil(sqrt(area_3d / area_2d) / resolution));
+      break;
+    }
+  }
+  for (auto& tc : tcoords)
+  {
+    tc.y() = 1.0 - tc.y();
```

Comment:
why do we have to flip Y here?

Reply:
We need to flip Y because the texture coordinate (0, 0) is the bottom-left corner whereas the image origin is the top-left corner. Y is flipped back at the end of the function.